partiql-eval/src/main/kotlin/org/partiql/eval/internal/Compiler.kt

```diff
@@ -31,127 +32,130 @@ import org.partiql.plan.Statement
 import org.partiql.plan.visitor.PlanBaseVisitor
 import org.partiql.spi.connector.ConnectorBindings
 import org.partiql.spi.connector.ConnectorObjectPath
+import org.partiql.types.StaticType
 import org.partiql.value.PartiQLValueExperimental
 import java.lang.IllegalStateException
 
 internal class Compiler(
     private val plan: PartiQLPlan,
     private val catalogs: Map<String, ConnectorBindings>,
-) : PlanBaseVisitor<Operator, Unit>() {
+) : PlanBaseVisitor<Operator, StaticType?>() { // Current ctx is a StaticType to differentiate between collection types for Rex.Op.Collection
 
     fun compile(): Operator.Expr {
-        return visitPartiQLPlan(plan, Unit)
+        return visitPartiQLPlan(plan, null)
     }
 
-    override fun defaultReturn(node: PlanNode, ctx: Unit): Operator {
+    override fun defaultReturn(node: PlanNode, ctx: StaticType?): Operator {
         TODO("Not yet implemented")
     }
 
-    override fun visitRexOpErr(node: Rex.Op.Err, ctx: Unit): Operator {
+    override fun visitRexOpErr(node: Rex.Op.Err, ctx: StaticType?): Operator {
         throw IllegalStateException(node.message)
     }
 
-    override fun visitRelOpErr(node: Rel.Op.Err, ctx: Unit): Operator {
+    override fun visitRelOpErr(node: Rel.Op.Err, ctx: StaticType?): Operator {
         throw IllegalStateException(node.message)
     }
 
     // TODO: Re-look at
-    override fun visitPartiQLPlan(node: PartiQLPlan, ctx: Unit): Operator.Expr {
+    override fun visitPartiQLPlan(node: PartiQLPlan, ctx: StaticType?): Operator.Expr {
         return visitStatement(node.statement, ctx) as Operator.Expr
     }
 
     // TODO: Re-look at
-    override fun visitStatementQuery(node: Statement.Query, ctx: Unit): Operator.Expr {
+    override fun visitStatementQuery(node: Statement.Query, ctx: StaticType?): Operator.Expr {
         return visitRex(node.root, ctx)
     }
 
     // REX
 
-    override fun visitRex(node: Rex, ctx: Unit): Operator.Expr {
-        return super.visitRexOp(node.op, ctx) as Operator.Expr
+    override fun visitRex(node: Rex, ctx: StaticType?): Operator.Expr {
```

Comment:
In the current plan, the rex node's type is only accessible on the `Rex` and not the inner classes (e.g. `Rex.Op.Collection`). Passing it along in the context allows the visitor for the inner classes to access the type, which is needed by some nodes like `Rex.Op.Collection`.

Reply:
Hmm I'd need to play around with the generated code a bit to see if there's a better way. Only concern I'd have is that the `type` field for `Rex` is a different type (i.e. `StaticType`) than the `type` field for `Rel` (i.e. `Rel.Type` which has schema and props). Probably that representation in the `PlanNode` base class would be an enum that could be either of those `type` definitions?

Anyways, perhaps we should tackle this in another issue/PR.

Reply:
> If you check out PlanTyper, you'll see we use two different visitors since the type is parameterized.
Oh I see. The separate visitors for Rel and Rex doesn't seem too cumbersome.

> The best situation however would be handwritten nodes with annotation based generation like Lombok. This would give us the most control.
Agree w/ a mix of code-generated nodes and handwritten nodes would give us the most flexibility when it comes to these interfaces.